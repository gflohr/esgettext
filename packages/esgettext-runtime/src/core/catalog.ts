export interface CatalogEntries {
	[key: string]: Array<string>;
}

export interface Catalog {
	major: number;
	minor: number;
	pluralFunction: Function;
	entries: CatalogEntries;
}
