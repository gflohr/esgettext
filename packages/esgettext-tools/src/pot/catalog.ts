import { POTEntry } from './entry';

interface Cache {
	// msgid
	[key: string]: {
		// msgctxt
		[key: string]: POTEntry;
	};
}

export interface CatalogProperties {
	package?: string;
	version?: string;
	msgidBugsAddress?: string;
	foreignUser?: boolean;
	date?: string;
	sortOutput?: boolean;
	sortByFile?: boolean;
	noHeader?: boolean;
}

interface RenderOptions {
	width?: number;
}

/**
 * An entire message catalog that is subsequently filled.
 */
export class Catalog {
	private readonly cache: Cache = {};
	entries: Array<POTEntry>;

	constructor(readonly properties: CatalogProperties = {}) {
		this.entries = new Array<POTEntry>();

		if (!properties.noHeader) {
			if (typeof properties.package === 'undefined') {
				properties.package = 'PACKAGE';
			}
			if (typeof properties.version === 'undefined') {
				properties.version = 'VERSION';
			}
			if (typeof properties.msgidBugsAddress === 'undefined') {
				properties.msgidBugsAddress = 'MSGID_BUGS_ADDRESS';
			}
			if (typeof properties.date === 'undefined') {
				const now = new Date();

				// Avoid if/else, so we do not spoil our test coverage. :)
				let year = now.getFullYear().toString();
				year = '0'.repeat(4 - year.length) + year;
				let month = (1 + now.getMonth()).toString();
				month = '0'.repeat(2 - month.length) + month;
				let mday = now.getDate().toString();
				mday = '0'.repeat(2 - mday.length) + mday;
				let hour = now.getHours().toString();
				hour = '0'.repeat(2 - hour.length) + hour;
				let minutes = now.getMinutes().toString();
				minutes = '0'.repeat(2 - minutes.length) + minutes;

				// Do not depend on the timezone for test coverage.
				let offset = now.getTimezoneOffset();
				const sign = ['+', '-'][Math.sign(offset) + 1];
				offset = Math.abs(offset);

				let offsetHours = Math.floor(offset / 60).toString();
				offsetHours = '0'.repeat(2 - offsetHours.length) + offsetHours;
				let offsetMinutes = (offset % 60).toString();
				offsetMinutes = '0'.repeat(2 - offsetMinutes.length) + offsetMinutes;

				properties.date =
					[year, month, mday].join('-') +
					` ${hour}:${minutes}` +
					`${sign}${offsetHours}${offsetMinutes}`;
			}

			const pkg = properties.package.replace(/\n/g, '\\n');
			const version = properties.version.replace(/\n/g, '\\n');
			const msgidBugsAddress = properties.msgidBugsAddress.replace(
				/\n/g,
				'\\n',
			);

			const header = `Project-Id-Version: ${pkg} ${version}
Report-Msgid-Bugs-To: ${msgidBugsAddress}
POT-Creation-Date: ${properties.date}
PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE
Last-Translator: FULL NAME <EMAIL@ADDRESS>
Language-Team: LANGUAGE <LL@li.org>
Language:
MIME-Version: 1.0
Content-Type: text/plain; charset=CHARSET
Content-Transfer-Encoding: 8bit
`;

			let comment;
			if (properties.foreignUser) {
				comment = `SOME DESCRIPTIVE TITLE
This file is put in the public domain.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
			} else {
				comment = `SOME DESCRIPTIVE TITLE
Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
This file is distributed under the same license as the PACKAGE package.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
			}

			const headerEntry = new POTEntry({
				msgid: '',
				msgstr: header,
				flags: ['fuzzy'],
				translatorComments: [comment],
				noWarnings: true,
			});
			this.addEntry(headerEntry);
		}
	}

	/**
	 * Render the entire catalog.
	 *
	 * @param renderOptions - options for rendering
	 *
	 * The following options are supported:
	 *
	 * width: wrap lines to that page width (default: 76)
	 *
	 * @returns the catalog rendered as a po(t) file
	 */
	toString(options?: RenderOptions): string {
		const width =
			options && typeof options.width !== 'undefined' ? options.width : 76;

		const isNotHeader = (entry: POTEntry): boolean => {
			return !(
				entry.properties.msgid === '' &&
				typeof entry.properties.msgctxt === 'undefined'
			);
		};
		const isHeader = (entry: POTEntry): boolean => {
			return !isNotHeader(entry);
		};

		const header = this.entries
			.filter(isHeader)
			.map(entry => entry.toString(width))
			.join('\n');

		let body = '';
		if (this.properties.sortOutput) {
			body = this.entries
				.filter(isNotHeader)
				.sort((a, b) => {
					const cmp = a.properties.msgid.localeCompare(b.properties.msgid);
					if (cmp) {
						return cmp;
					}
					const actx = a.properties.msgctxt;
					const bctx = b.properties.msgctxt;

					// It is impossible that both msgctxts are undefined
					// because that would mean that the msgids are identical
					// which is prevented.
					if (typeof bctx === 'undefined') {
						return +1;
					} else if (typeof actx === 'undefined') {
						return -1;
					}

					return actx.localeCompare(bctx);
				})
				.map(entry => entry.toString(width))
				.join('\n');
		} else if (this.properties.sortByFile) {
			const copy = this.copy(this.properties);

			const splitRef = (ref: string): { filename: string; lineno: number } => {
				const parts = ref.split(':');
				const lineno = Number.parseInt(parts.pop(), 10);
				const filename = parts.join(':');
				return { filename, lineno };
			};

			// First order all references
			copy.entries.forEach(entry => {
				if (entry.properties.references) {
					entry.properties.references = entry.properties.references
						.map(splitRef)
						.sort((a, b) => {
							return (
								a.filename.localeCompare(b.filename) ||
								Math.sign(a.lineno - b.lineno)
							);
						})
						.map(ref => `${ref.filename}:${ref.lineno}`);
				}
			});

			const compareSplitRefs = (
				a: {
					refs: Array<{ filename: string; lineno: number }>;
					entry: POTEntry;
				},
				b: {
					refs: Array<{ filename: string; lineno: number }>;
					entry: POTEntry;
				},
			): number => {
				const arefs = a.refs;
				const brefs = b.refs;
				const min = Math.min(arefs.length, brefs.length);
				for (let i = 0; i < min; ++i) {
					const cmp =
						arefs[i].filename.localeCompare(brefs[i].filename) ||
						Math.sign(arefs[i].lineno - brefs[i].lineno);
					if (cmp) {
						return cmp;
					}
				}
				return Math.sign(arefs.length - brefs.length);
			};

			const splitRefs = (
				refs: Array<string>,
			): Array<{ filename: string; lineno: number }> => {
				return refs ? refs.map(splitRef) : [];
			};

			body = copy.entries
				.filter(isNotHeader)
				.map(entry => {
					return { refs: splitRefs(entry.properties.references), entry };
				})
				.sort(compareSplitRefs)
				.map(split => split.entry)
				.map(entry => entry.toString(width))
				.join('\n');
		} else {
			body = this.entries
				.filter(isNotHeader)
				.map(entry => entry.toString(width))
				.join('\n');
		}

		if (body && header) {
			return `${header}\n${body}`;
		} else {
			return header || body || '';
		}
	}

	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `POTEntry` to add
	 */
	addEntry(entry: POTEntry): void {
		const msgid = entry.properties.msgid;
		const msgctxt = entry.properties.msgctxt;

		if (!Object.prototype.hasOwnProperty.call(this.cache, msgid)) {
			this.cache[msgid] = {};
		}
		if (!Object.prototype.hasOwnProperty.call(this.cache[msgid], msgctxt)) {
			this.cache[msgid][msgctxt] = entry;
			this.entries.push(entry);
		} else {
			this.cache[msgid][msgctxt].merge(entry);
		}
	}

	deleteEntry(msgid: string, msgctxt?: string): void {
		if (!this.cache[msgid]) {
			return;
		}

		const entry = this.cache[msgid][msgctxt];
		if (!entry) {
			return;
		}

		delete this.cache[msgid][msgctxt];
		this.entries = this.entries.filter(other => other !== entry);
	}

	/**
	 * Copy a catalog with other options. This is for testing only.
	 *
	 * @returns a deep copy of the catalog.
	 */
	copy(properties?: CatalogProperties): Catalog {
		const other = new Catalog(properties);

		this.entries.forEach(entry => other.addEntry(entry));

		return other;
	}

	/**
	 * Removes all translations from a catalog, so that it can be used as
	 * a .pot file.
	 */
	makePOT(): void {
		this.entries.forEach(entry => {
			if (
				!(
					entry.properties.msgid === '' &&
					typeof entry.properties.msgctxt === 'undefined'
				)
			) {
				delete entry.properties.msgstr;
			}
		});
	}
}
