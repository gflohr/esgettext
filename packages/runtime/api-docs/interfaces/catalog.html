<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>Catalog | @esgettext/runtime</title>
	<meta name="description" content="Documentation for @esgettext/runtime">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="../assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.json" data-base="..">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="../index.html" class="title">@esgettext/runtime</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="../globals.html">Globals</a>
				</li>
				<li>
					<a href="catalog.html">Catalog</a>
				</li>
			</ul>
			<h1>Interface Catalog</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<section class="tsd-panel tsd-comment">
				<div class="tsd-comment tsd-typography">
					<div class="lead">
						<p>A <a href="catalog.html"><code>Catalog</code></a> is a container for a set of translations loaded from a <code>.json</code>
						or a binary <code>.mo</code> file.</p>
					</div>
					<p>This interface is used internally. You will only need it if you want to
						write your own message retrieval method or want to inspect a loaded
					<a href="catalog.html"><code>Catalog</code></a>.</p>
				</div>
			</section>
			<section class="tsd-panel tsd-hierarchy">
				<h3>Hierarchy</h3>
				<ul class="tsd-hierarchy">
					<li>
						<span class="target">Catalog</span>
					</li>
				</ul>
			</section>
			<section class="tsd-panel-group tsd-index-group">
				<h2>Index</h2>
				<section class="tsd-panel tsd-index-panel">
					<div class="tsd-index-content">
						<section class="tsd-index-section ">
							<h3>Properties</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-property tsd-parent-kind-interface"><a href="catalog.html#entries" class="tsd-kind-icon">entries</a></li>
								<li class="tsd-kind-property tsd-parent-kind-interface"><a href="catalog.html#major" class="tsd-kind-icon">major</a></li>
								<li class="tsd-kind-property tsd-parent-kind-interface"><a href="catalog.html#minor" class="tsd-kind-icon">minor</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Methods</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-method tsd-parent-kind-interface"><a href="catalog.html#pluralfunction" class="tsd-kind-icon">plural<wbr>Function</a></li>
							</ul>
						</section>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Properties</h2>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-interface">
					<a name="entries" class="tsd-anchor"></a>
					<h3>entries</h3>
					<div class="tsd-signature tsd-kind-icon">entries<span class="tsd-signature-symbol">:</span> <a href="catalogentries.html" class="tsd-signature-type">CatalogEntries</a></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gflohr/esgettext/blob/2738702/packages/runtime/src/core/catalog.ts#L50">src/core/catalog.ts:50</a></li>
						</ul>
					</aside>
					<div class="tsd-comment tsd-typography">
						<div class="lead">
							<p>The actual translations.</p>
						</div>
					</div>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-interface">
					<a name="major" class="tsd-anchor"></a>
					<h3>major</h3>
					<div class="tsd-signature tsd-kind-icon">major<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">number</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gflohr/esgettext/blob/2738702/packages/runtime/src/core/catalog.ts#L30">src/core/catalog.ts:30</a></li>
						</ul>
					</aside>
					<div class="tsd-comment tsd-typography">
						<div class="lead">
							<p>The major revision number of the catalog, currently always 0.</p>
						</div>
					</div>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-interface">
					<a name="minor" class="tsd-anchor"></a>
					<h3>minor</h3>
					<div class="tsd-signature tsd-kind-icon">minor<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">number</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gflohr/esgettext/blob/2738702/packages/runtime/src/core/catalog.ts#L33">src/core/catalog.ts:33</a></li>
						</ul>
					</aside>
					<div class="tsd-comment tsd-typography">
						<div class="lead">
							<p>The minor revision number of the catalog, currently always 0 or 1.</p>
						</div>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Methods</h2>
				<section class="tsd-panel tsd-member tsd-kind-method tsd-parent-kind-interface">
					<a name="pluralfunction" class="tsd-anchor"></a>
					<h3>plural<wbr>Function</h3>
					<ul class="tsd-signatures tsd-kind-method tsd-parent-kind-interface">
						<li class="tsd-signature tsd-kind-icon">plural<wbr>Function<span class="tsd-signature-symbol">(</span>numItems<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">number</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/gflohr/esgettext/blob/2738702/packages/runtime/src/core/catalog.ts#L47">src/core/catalog.ts:47</a></li>
								</ul>
							</aside>
							<div class="tsd-comment tsd-typography">
								<div class="lead">
									<p>Compute the index of a plural form from a number of items. If a language
										has one singular and two plural forms, the singular form would have
									index 0 and the plural forms have index 1 and 2.</p>
								</div>
								<p>The plural function would then compute one of 0, 1, or 2 from an
								arbitrary non-negative integer.</p>
							</div>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>numItems: <span class="tsd-signature-type">number</span></h5>
									<div class="tsd-comment tsd-typography">
										<p>the number of items.</p>
									</div>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">number</span></h4>
							<p>the index of the plural form.</p>
						</li>
					</ul>
				</section>
			</section>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="../globals.html"><em>Globals</em></a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-class">
						<a href="../classes/textdomain.html" class="tsd-kind-icon">Textdomain</a>
					</li>
				</ul>
				<ul class="current">
					<li class="current tsd-kind-interface">
						<a href="catalog.html" class="tsd-kind-icon">Catalog</a>
						<ul>
							<li class=" tsd-kind-property tsd-parent-kind-interface">
								<a href="catalog.html#entries" class="tsd-kind-icon">entries</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-interface">
								<a href="catalog.html#major" class="tsd-kind-icon">major</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-interface">
								<a href="catalog.html#minor" class="tsd-kind-icon">minor</a>
							</li>
							<li class=" tsd-kind-method tsd-parent-kind-interface">
								<a href="catalog.html#pluralfunction" class="tsd-kind-icon">plural<wbr>Function</a>
							</li>
						</ul>
					</li>
				</ul>
				<ul class="after-current">
					<li class=" tsd-kind-interface">
						<a href="catalogentries.html" class="tsd-kind-icon">Catalog<wbr>Entries</a>
					</li>
					<li class=" tsd-kind-interface">
						<a href="localecontainer.html" class="tsd-kind-icon">Locale<wbr>Container</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-class"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-class"><span class="tsd-kind-icon">Method</span></li>
				<li class="tsd-kind-accessor tsd-parent-kind-class"><span class="tsd-kind-icon">Accessor</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-interface"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-interface"><span class="tsd-kind-icon">Method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-method tsd-parent-kind-class tsd-is-static"><span class="tsd-kind-icon">Static method</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="../assets/js/main.js"></script>
</body>
</html>