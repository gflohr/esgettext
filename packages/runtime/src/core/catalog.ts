/**
 * The set of translations found in a [[`Catalog`]].
 *
 * This interface is used internally. You will only need it if you want to
 * write your own message retrieval method or want to inspect a loaded
 * [[`Catalog`]].
 *
 * The translations are looked up by their singular form in the original
 * language. The value is an array of strings. The first item is the singular
 * translation, the optional following items are the plural forms.
 *
 * If a translation has a message context, the key is the context joined
 * with the translation by a `'\u0004'` character. For example, the key for
 * the msgid "Open" with the msgctxt "Menu|File" would be `Menu|File\u0004Open`.
 */
export interface CatalogEntries {
	[key: string]: Array<string>;
}

/**
 * A [[`Catalog`]] is a container for a set of translations loaded from a
 * `mo.json` or a binary `.mo` file.
 *
 * This interface is used internally. You will only need it if you want to
 * write your own message retrieval method or want to inspect a loaded
 * [[`Catalog`]].
 */
export interface Catalog {
	/** The major revision number of the catalog, currently always 0. */
	major: number;

	/** The minor revision number of the catalog, currently always 0 or 1. */
	minor: number;

	/**
	 * Compute the index of a plural form from a number of items. If a language
	 * has one singular and two plural forms, the singular form would have
	 * index 0 and the plural forms have index 1 and 2.
	 *
	 * The plural function would then compute one of 0, 1, or 2 from an
	 * arbitrary non-negative integer.
	 *
	 * @param numItems - the number of items.
	 *
	 * @returns the index of the plural form.
	 */
	pluralFunction(numItems: number): number;

	/** The actual translations. */
	entries: CatalogEntries;
}
