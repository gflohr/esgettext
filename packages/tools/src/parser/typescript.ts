import { parse } from '@babel/parser';
import { Parser } from './parser';

export class TypeScriptParser extends Parser {
	doParse(input: string, filename: string): boolean {
		const ast = parse(input, {
			allowAwaitOutsideFunction: true,
			allowImportExportEverywhere: true,
			allowReturnOutsideFunction: true,
			allowSuperOutsideMethod: true,
			allowUndeclaredExports: true,
			errorRecovery: true,
			sourceFilename: filename,
			sourceType: this.getSourceType(filename),
			plugins: [
				'typescript',
				'jsx',
				'classProperties',
				'classPrivateMethods',
				'classPrivateProperties',
				'decorators',
			],
		});

		return this.extract(filename, ast);
	}
}
