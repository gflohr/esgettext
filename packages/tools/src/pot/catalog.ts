import { POTEntry } from './entry';

interface Cache {
	// msgid
	[key: string]: {
		// msgctxt
		[key: string]: POTEntry;
	};
}

interface CatalogProperties {
	package?: string;
	version?: string;
	msgidBugsAddress?: string;
	foreignUser?: boolean;
	date?: string;
	fromCode?: string;
	sortOutput?: boolean;
	sortByFile?: boolean;
}

interface RenderOptions {
	width?: number;
}

/**
 * An entire message catalog that is subsequently filled.
 */
export class Catalog {
	private readonly cache: Cache = {};
	private readonly entries: Array<POTEntry>;

	constructor(private readonly properties: CatalogProperties = {}) {
		if (typeof properties.package === 'undefined') {
			properties.package = 'PACKAGE';
		}
		if (typeof properties.version === 'undefined') {
			properties.version = 'VERSION';
		}
		if (typeof properties.msgidBugsAddress === 'undefined') {
			properties.msgidBugsAddress = 'MSGID_BUGS_ADDRESS';
		}
		// GNU xgettext does not set the charset in the PO header but we do.
		if (typeof properties.fromCode === 'undefined') {
			properties.fromCode = 'CHARSET';
		}
		if (typeof properties.date === 'undefined') {
			const now = new Date();

			// Avoid if/else, so we do not spoil our test coverage. :)
			let year = now.getFullYear().toString();
			year = '0'.repeat(4 - year.length) + year;
			let month = now.getMonth().toString();
			month = '0'.repeat(2 - month.length) + month;
			let mday = now.getDate().toString();
			mday = '0'.repeat(2 - mday.length) + mday;
			let hour = now.getHours().toString();
			hour = '0'.repeat(2 - hour.length) + hour;
			let minutes = now.getMinutes().toString();
			minutes = '0'.repeat(2 - minutes.length) + minutes;

			// Do not depend on the timezone for test coverage.
			let offset = now.getTimezoneOffset();
			const sign = ['+', '-'][Math.sign(offset) + 1];
			offset = Math.abs(offset);

			let offsetHours = Math.floor(offset / 60).toString();
			offsetHours = '0'.repeat(2 - offsetHours.length) + offsetHours;
			let offsetMinutes = (offset % 60).toString();
			offsetMinutes = '0'.repeat(2 - offsetMinutes.length) + offsetMinutes;

			properties.date =
				[year, month, mday].join('-') +
				` ${hour}:${minutes}` +
				`${sign}${offsetHours}${offsetMinutes}`;
		}

		const pkg = properties.package.replace(/\n/g, '\\n');
		const version = properties.version.replace(/\n/g, '\\n');
		const msgidBugsAddress = properties.msgidBugsAddress.replace(/\n/g, '\\n');

		const header = `Project-Id-Version: ${pkg} ${version}
Report-Msgid-Bugs-To: ${msgidBugsAddress}
POT-Creation-Date: ${properties.date}
PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE
Last-Translator: FULL NAME <EMAIL@ADDRESS>
Language-Team: LANGUAGE <LL@li.org>
Language:
MIME-Version: 1.0
Content-Type: text/plain; charset=${properties.fromCode}
Content-Transfer-Encoding: 8bit
`;

		let comment;
		if (properties.foreignUser) {
			comment = `SOME DESCRIPTIVE TITLE
This file is put in the public domain.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		} else {
			comment = `SOME DESCRIPTIVE TITLE
Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
This file is distributed under the same license as the PACKAGE package.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		}

		this.entries = new Array<POTEntry>();
		const headerEntry = new POTEntry({
			msgid: '',
			msgstr: header,
			flags: ['fuzzy'],
			translatorComments: [comment],
			noWarnings: true,
		});
		this.addEntry(headerEntry);
	}

	/**
	 * Render the entire catalog.
	 *
	 * @param renderOptions - options for rendering
	 *
	 * The following options are supported:
	 *
	 * width: wrap lines to that page width (default: 76)
	 *
	 * @returns the catalog rendered as a po(t) file
	 */
	toString(options?: RenderOptions): string {
		const width =
			options && typeof options.width !== 'undefined' ? options.width : 76;

		if (this.properties.sortOutput) {
			return this.entries
				.sort((a, b) => {
					const cmp = a.properties.msgid.localeCompare(b.properties.msgid);
					if (cmp) {
						return cmp;
					}
					const actx = a.properties.msgctxt;
					const bctx = b.properties.msgctxt;

					if (typeof actx === 'undefined') {
						if (typeof bctx === 'undefined') {
							return 0;
						} else {
							return -1;
						}
					} else {
						if (typeof bctx === 'undefined') {
							return +1;
						} else {
							return actx.localeCompare(bctx);
						}
					}
				})
				.map((entry) => entry.toString(width))
				.join('\n');
		} else {
			return this.entries.map((entry) => entry.toString(width)).join('\n');
		}
	}

	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `POTEntry` to add
	 */
	addEntry(entry: POTEntry): void {
		const msgid = entry.properties.msgid;
		const msgctxt = entry.properties.msgctxt;

		if (!Object.prototype.hasOwnProperty.call(this.cache, msgid)) {
			this.cache[msgid] = {};
		}
		if (!Object.prototype.hasOwnProperty.call(this.cache[msgid], msgctxt)) {
			this.cache[msgid][msgctxt] = entry;
			this.entries.push(entry);
		} else {
			this.cache[msgid][msgctxt].merge(entry);
		}
	}

	/**
	 * Copy a catalog with other options. This is for testing only.
	 *
	 * @returns a deep copy of the catalog.
	 */
	copy(properties?: CatalogProperties): Catalog {
		const other = new Catalog(properties);

		this.entries.forEach((entry) => other.addEntry(entry));

		return other;
	}
}
