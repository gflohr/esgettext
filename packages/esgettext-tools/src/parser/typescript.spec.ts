import { Catalog } from '../pot/catalog';
import { TypeScriptParser } from './typescript';

describe('TypeScript parser', () => {
	describe('strings', () => {
		it('should parse a simple call', () => {
			const catalog = new Catalog({ noHeader: true });
			const warner = jest.fn();
			const p = new TypeScriptParser(catalog, warner);
			const code = 'gtx._("Hello, world!")';
			expect(p.parse(Buffer.from(code), 'example.ts')).toBeTruthy();
			const expected = `#: example.ts:1
msgid "Hello, world!"
msgstr ""
`;
			expect(catalog.toString()).toEqual(expected);
			expect(warner).not.toHaveBeenCalled();
		});

		it('should parse itself', () => {
			const catalog = new Catalog({ noHeader: true });
			const warner = jest.fn();
			const p = new TypeScriptParser(catalog, warner);
			const filename = __filename.replace(/typescript\.spec\.ts$/, 'parser.ts');
			expect(p.parseFile(filename)).toBeTruthy();
			expect(catalog.toString()).toMatchSnapshot();
		});

		it('should parse itself with encoding', () => {
			const catalog = new Catalog({ noHeader: true });
			const warner = jest.fn();
			const p = new TypeScriptParser(catalog, warner);
			const filename = __filename.replace(/typescript\.spec\.ts$/, 'parser.ts');
			expect(p.parseFile(filename, 'utf-8')).toBeTruthy();
			expect(catalog.toString()).toMatchSnapshot();
		});
	});
});
