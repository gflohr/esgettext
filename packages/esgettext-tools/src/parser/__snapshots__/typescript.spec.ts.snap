// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript parser strings should parse itself 1`] = `
"#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:216
msgid \\"\\"
\\"template literals with embedded expressions are not allowed as arguments \\"
\\"to gettext functions because they are not constant\\"
msgstr \\"\\"

#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:403
#, perl-brace-format
msgid \\"{location}: warning: {msg}\\"
msgstr \\"\\"

#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:411
#, perl-brace-format
msgid \\"{location}: error: {msg}\\"
msgstr \\"\\"
"
`;

exports[`TypeScript parser strings should parse itself with encoding 1`] = `
"#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:216
msgid \\"\\"
\\"template literals with embedded expressions are not allowed as arguments \\"
\\"to gettext functions because they are not constant\\"
msgstr \\"\\"

#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:403
#, perl-brace-format
msgid \\"{location}: warning: {msg}\\"
msgstr \\"\\"

#: /Users/guido/javascript/esgettext/packages/esgettext-tools/src/parser/parser.ts:411
#, perl-brace-format
msgid \\"{location}: error: {msg}\\"
msgstr \\"\\"
"
`;
